StreamReader.cpp

```diff
@@ -39,18 +37,18 @@ MemoryStreamReader::MemoryStreamReader(char* buffer, size_t size) {
 void MemoryStreamReader::Read(char* buffer, size_t size) 
 {
 	if (offset + size > streamSize) {
-		throw runtime_error("Size of bytes to read exceeds remaining size of buffer.");
+		throw std::runtime_error("Size of bytes to read exceeds remaining size of buffer.");
 	}
 
 	memcpy(buffer, streamBuffer + offset, size);
 	offset += size;
 }
 
-void MemoryStreamReader::Ignore(size_t size)
+void MemoryStreamReader::Seek(int offset)
 {
-	if (offset + size > streamSize) {
-		throw runtime_error("Size of bytes to ignore exceeds remaining size of buffer.");
+	if (this->offset + offset > streamSize || this->offset + offset < 0) {
```

Comment:
The `size_t` type is unsigned. When you mix signed and unsigned types, the signed type is converted to unsigned. Hence the result of the addition is unsigned. An unsigned type can never be `< 0`. The comparison against `streamSize` should be enough to catch both overflow and underflow.

Reply:
So, in this case, would it set size_t this->offset to 0 if a subtraction is performed that would otherwise take a signed number into the negative range?

I think even if protection is provided from reading outside the stream, it would be nice to inform the user they are attempting to read outside the bounds as this is probably not their intention. Either way it is currently broken as you mentioned.

Thanks.

Reply:
Unsigned numbers have well defined wrap around. Instead of going negative, it will turn into a large positive, and be caught by the first condition.